core/src/com/unciv/logic/city/IConstruction.kt

```diff
@@ -1,23 +1,99 @@
+import com.unciv.models.stats.Stat
 import com.unciv.ui.utils.Fonts
+import kotlin.math.pow
 import kotlin.math.roundToInt
 
 interface IConstruction : INamed {
-    fun getProductionCost(civInfo: CivilizationInfo): Int
-    fun getGoldCost(civInfo: CivilizationInfo): Int
     fun isBuildable(cityConstructions: CityConstructions): Boolean
     fun shouldBeDisplayed(cityConstructions: CityConstructions): Boolean
     fun postBuildEvent(cityConstructions: CityConstructions, wasBought: Boolean = false): Boolean  // Yes I'm hilarious.
     fun getResourceRequirements(): HashMap<String,Int>
-    fun canBePurchased(): Boolean
+}
+
+interface INonPerpetualConstruction : IConstruction, INamed {
+    val hurryCostModifier: Int
+    val uniqueObjects: List<Unique>
+    val uniques: List<String>
+
+    fun getProductionCost(civInfo: CivilizationInfo): Int
+    fun getStatBuyCost(cityInfo: CityInfo, stat: Stat): Int?
+    
+    private fun getMatchingUniques(uniqueTemplate: String): Sequence<Unique> {
+        return uniqueObjects.asSequence().filter { it.placeholderText == uniqueTemplate }
+    }
+    
+    fun canBePurchasedWithStat(cityInfo: CityInfo, stat: Stat): Boolean {
+        if (stat in listOf(Stat.Production, Stat.Happiness)) return false
+        if ("Cannot be purchased" in uniques) return false
+        if (stat == Stat.Gold) return !uniques.contains("Unbuildable")
+        if (getMatchingUniques("Can be purchased with []")
+                .filter { it.params[0] == stat.name }
+                .any()
+        ) return true
+        if (getMatchingUniques("Can be purchased with [] []")
+                .filter { it.params[0] == stat.name && cityInfo.matchesFilter(it.params[1]) }
+                .any()
+        ) return true
+        if (getMatchingUniques("Can be purchased for [] []")
+                .filter { it.params[1] == stat.name}
+                .any()
+        ) return true
+        if (getMatchingUniques("Can be purchased for [] [] []")
+                .filter { it.params[1] == stat.name && cityInfo.matchesFilter(it.params[2]) }
+                .any()
+        ) return true
+        return false
+    }
+    
+    fun canBePurchasedWithAnyStat(cityInfo: CityInfo): Boolean {
+        return Stat.values().any { canBePurchasedWithStat(cityInfo, it) }
+    }
+    
+    // I can't make this function protected or private :(
+    fun getBaseGoldCost(civInfo: CivilizationInfo): Double {
+        // https://forums.civfanatics.com/threads/rush-buying-formula.393892/
+        return (30.0 * getProductionCost(civInfo)).pow(0.75) * (1 + hurryCostModifier / 100f)
+    }
+    
+    // I can't make this function protected or private :(
+    fun getBaseBuyCost(cityInfo: CityInfo, stat: Stat): Double? {
+        if (stat == Stat.Gold) return getBaseGoldCost(cityInfo.civInfo)
+
+        val cost = getMatchingUniques("Can be purchased for [] []")
+            .filter { it.params[1] == stat.name }
+            .minByOrNull { it.params[0].toInt() }
```

Comment:
Just !! this, to make it Int instead of Int?, and your problems will disappear :)

Reply:
I don't think I can do that, as `emptySequence().minByOrNull { [...] } == null`, so that would result in a nullptrexception when it construction has no such uniques, right?